src/aiida_quantumespresso/calculations/__init__.py

```diff
@@ -136,11 +136,29 @@ def define(cls, spec):
                     for flag_name in cls._ENABLED_PARALLELIZATION_FLAGS
                 )
             ),
-            validator=cls._validate_parallelization
+            validator=cls.validate_parallelization
         )
+        spec.inputs.validator = cls.validate_inputs
 
     @classmethod
-    def _validate_parallelization(cls, value, port_namespace):  # pylint: disable=unused-argument
+    def validate_inputs(cls, value, _):
+        """Validate the entire inputs namespace."""
+        if 'pseudos' not in value:
+            return 'required value was not provided for the `pseudos` namespace.'
+
+        # Some wrapping processes expose the inputs excluding the ``structure``, in which case we can't validate.
+        if 'structure' not in value:
+            return
+
+        structure_kinds = value['structure'].get_kind_names()
+        pseudo_kinds = list(value['pseudos'].keys())
+
+        if structure_kinds != pseudo_kinds:
```

Comment:
Should we sort the kinds list first, or use sets? They don't necessarily have the same order.

Reply:
Astute observation and you are absolutely correct. I used sets initially, but had some kind of error, so removed it. But we should use it yeah, will adapt.